src/blocks/calendar.rs

```diff
@@ -0,0 +1,396 @@
+        }
+    }
+}
+
+async fn open_browser(config: &Config, url: &Url) -> Result<()> {
+    let cmd = config.browser_cmd.expand()?;
+    has_command(&cmd)
+        .await
+        .or_error(|| "Browser command not found")?;
+    spawn_process(&cmd, &[url.as_ref()]).error("Open browser failed")
+}
+
+async fn caldav_client(config: &Config) -> Result<caldav::CalDavClient> {
+    let auth = match &config.auth {
+        AuthConfig::Unauthenticated => auth::Auth::Unauthenticated,
+        AuthConfig::Basic(BasicAuthConfig { username, password }) => {
+            let username = username
+                .clone()
+                .or_else(|| std::env::var("I3RS_CALENDAR_AUTH_USERNAME").ok())
+                .error("Calendar username not found")?;
+            let password = password
+                .clone()
+                .or_else(|| std::env::var("I3RS_CALENDAR_AUTH_PASSWORD").ok())
+                .error("Calendar password not found")?;
+            auth::Auth::basic(username, password)
+        }
+        AuthConfig::OAuth2(oauth2) => {
+            let client_id = oauth2
+                .client_id
+                .clone()
+                .or_else(|| std::env::var("I3RS_CALENDAR_AUTH_CLIENT_ID").ok())
+                .error("Calendar oauth2 client_id not found")?;
+            let client_secret = oauth2
+                .client_secret
+                .clone()
+                .or_else(|| std::env::var("I3RS_CALENDAR_AUTH_CLIENT_SECRET").ok())
+                .error("Calendar oauth2 client_secret not found")?;
+            let auth_url =
+                AuthUrl::new(oauth2.auth_url.clone()).error("Invalid authorization url")?;
+            let token_url = TokenUrl::new(oauth2.token_url.clone()).error("Invalid token url")?;
+
+            let flow = OAuth2Flow::new(
+                ClientId::new(client_id),
+                ClientSecret::new(client_secret),
+                auth_url,
+                token_url,
+                oauth2.redirect_port,
+            );
+            let token_store = TokenStore::new(Path::new(&oauth2.auth_token.expand()?.to_string()));
+            auth::Auth::oauth2(flow, token_store, oauth2.scopes.clone())
+        }
+    };
+    Ok(CalDavClient::new(
+        Url::parse(&config.url).error("Invalid CalDav server url")?,
+        auth,
+    ))
+}
+
+async fn get_next_event(
+    config: &Config,
+    client: &mut CalDavClient,
+) -> Result<Option<caldav::Event>, CalendarError> {
+    let calendars: Vec<_> = client
+        .calendars()
+        .await?
+        .into_iter()
+        .filter(|c| config.calendars.is_empty() || config.calendars.contains(&c.name))
+        .collect();
+    let mut events: Vec<Event> = vec![];
+    for calendar in calendars {
+        let calendar_events: Vec<_> = client
+            .events(
+                &calendar,
+                Utc::now(),
+                Utc::now()
+                    + Duration::try_days(config.events_within_hours.try_into().unwrap()).unwrap(),
```

Comment:
This should be `try_hours`.  I'd make `events_within_hours` `u32` to constrain it to positive numbers that will fit within the i64 that `try_hours` takes.